### DISCUSSION TITLE
Localization & Translations

### DISCUSSION THREAD

**Discussion by ssandino** [socialincome-san/public#57] [Archived Discussions]
**How can we streamline the localization of the website, communication and the mobile app?** 

Some further thoughts:
- How to invite the open source community to participate in the translations?
- Edit text copy without coding? 
- Integrating with GitHub (developers approve the merge request)

**Comment by ssandino** [comment]
[Crowdin](https://crowdin.com) seems an option: "It’s free for [open source projects](https://crowdin.com/page/open-source-project-setup-request), it integrates well with GitHub, and has moderation controls for managing large translator communities." 

It's also used on [opencollective.com](https://opencollective.com/) in a interesting way: the official language switcher shows directly the amount of missing translations: 

![Screen Shot 2022-09-11 at 14 32 32](https://user-images.githubusercontent.com/6095849/189527695-3401b661-d775-44c2-a72b-9c951a8ffdd7.png)

**Comment by ssandino** [comment]
Another tool mentioned by @andrashee is [locize.com](https://locize.com)

**Comment by ssandino** [comment]
Some thoughts on using our existing setup with `FireCMS` to manage translations (inspired by @mkue who brought the topic up to host website content). 

Why would we want to use a third party tool to manage translations? Following reasons come to mind:

1. provides role-based account management (translators only see translation files)
2. allows non-developers from the open source community to contribute (specifically translators)
3. supports the human translations with Neural Machine Translation (NMT) (it proposes the best match)

These features could be (with some effort and to some extend) also provided by our existing admin tool: 

1. we have a secure user account management in place with role-based access to specific firebase collections
2. it is user-friendly for non-developers (and mobile friendly)
3. AI translations could be added with a simple [firebase extension](https://firebase.google.com/products/extensions/firebase-firestore-translate-text)

One challenge: if have all text strings for the website  as `fields` in `document` in  firebase `collection` we would have an incredible amount of requests compared to loading the current JSON file per language and page. BUT with a Google Cloud function that creates (triggered by a change) a JSON file and saves it on the cloud, we could work this out?

Just saying ... any thoughts @mkue, @andrashee, @sdwivedi?

**Comment by andrashee** [comment]
Interesting idea!

Generally, I would be an advocate of using an i8n framework for the translation.

They already solved all the problems of language detection, plurals, number, currency date formatting for us.

One which seems to be often used is e.g https://www.i18next.com.

It also comes with a next.js specific plugin: https://github.com/i18next/next-i18next
. Side note: next.js itself also has imho really good built in support to build international websites: https://nextjs.org/docs/advanced-features/i18n-routing.

These frameworks usually support quite some different input formats for the translations. JSON, getext, po files, ...

Creating for example the json files from firestore (curated through firecms) at build time of the website could be an option. I definitely wouldn't grab the individual translations from the database at serve time.

I personally would still explore the dedicated services for this, rather than using firecms. 

Locize is the company behind the framework mentioned above, but there are quite some of them.

https://localizely.com/blog/nextjs-i18n-tutorial/ is another one with good integration and free for open source.

Those translation management systems are explicitly built for this. E.g one can immediately see which keys are used in the code, but are missing in the translation and the other way around. Translations which are not referenced anymore in the code. 

This list could be a good inspiration https://github.com/jpomykala/awesome-i18n

**Comment by andrashee** [comment]
One more interesting link how locize is used in a next.js setup, its challenges with client and server side rendering and how to automatically sync code with the platform https://locize.com/blog/next-i18next/

**Comment by MDikkii** [comment]
From what I can see on the flutter packages localizely seems promising, it has also OTA updates as the only one from those mentioned, and as far as I can see has also integration with most used `intl` package for flutter. 

All other have some packages as well. And even if something won't work with the flutter packages we can still try to wrap it with our own code.

I haven't got strong opinion.

**Comment by ssandino** [comment]
i8n framework already in use.
